runtime/j9vm/javanextvmi.c

```diff
@@ -32,10 +36,87 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
+	J9ROMClass *romClass;
+	J9ROMMethod *romMethod;
+	UDATA bytecodeOffset;
+} GetStackTraceElementUserData;
+static UDATA
+getStackTraceElementIterator(J9VMThread *vmThread, void *voidUserData, UDATA bytecodeOffset, J9ROMClass *romClass, J9ROMMethod *romMethod, J9UTF8 *fileName, UDATA lineNumber, J9ClassLoader *classLoader)
+{
+	GetStackTraceElementUserData *userData = voidUserData;
+
+	/* We are done, only first stack frame is needed. */
+	userData->romClass = romClass;
+	userData->romMethod = romMethod;
+	userData->bytecodeOffset = bytecodeOffset;
+
+	return FALSE;
+}
+static void cfdumpBytecodePrintFunction(void *userData, char *format, ...)
+{
+	PORT_ACCESS_FROM_PORT((J9PortLibrary*)userData);
+	va_list args;
+	char outputBuffer[512] = {0};
+
+	va_start(args, format);
+	j9str_vprintf(outputBuffer, 512, format, args);
+	va_end(args);
+	j9tty_printf(PORTLIB, "%s", outputBuffer);
+}
+
 JNIEXPORT jstring JNICALL
 JVM_GetExtendedNPEMessage(JNIEnv *env, jthrowable throwableObj)
 {
-	/* Returning NULL to allow JDK14 compilation, https://github.com/eclipse/openj9/issues/7500 */
-	return NULL;
+	J9VMThread *vmThread = (J9VMThread *)env;
+	J9JavaVM *vm = vmThread->javaVM;
+	jobject msgObjectRef = NULL;
+	
+	Trc_SC_GetExtendedNPEMessage_Entry(vmThread, throwableObj);
+	if (J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_SHOW_EXTENDED_NPEMSG)) {
+		J9InternalVMFunctions const * const vmFuncs = vm->internalVMFunctions;
+		char *npeMsg = NULL;
+		GetStackTraceElementUserData userData = {0}; /* only first frame is needed */
+		
+		Trc_SC_GetExtendedNPEMessage_Entry2(vmThread, throwableObj);
+		vmFuncs->internalEnterVMFromJNI(vmThread);
+		userData.bytecodeOffset = UDATA_MAX;
+		vmFuncs->iterateStackTrace(vmThread, (j9object_t*)throwableObj, getStackTraceElementIterator, &userData, FALSE);
+		if ((NULL != userData.romClass) && (NULL != userData.romMethod) && (UDATA_MAX != userData.bytecodeOffset)) {
+			J9BytecodeVerificationData *bcvd = vm->bytecodeVerificationData;
+
+			if (TrcEnabled_Trc_SC_GetExtendedNPEMessage_Dump_Bytecode) {
+				PORT_ACCESS_FROM_VMC(vmThread);
+				U_8 *bytecodes = J9_BYTECODE_START_FROM_ROM_METHOD(userData.romMethod);
+				U_32 flags = 0;
+
+#ifdef J9VM_ENV_LITTLE_ENDIAN
+				flags |= BCT_LittleEndianOutput;
+#else
+				flags |= BCT_BigEndianOutput;
+#endif
+				j9bcutil_dumpBytecodes(PORTLIB, userData.romClass, bytecodes, 0, userData.bytecodeOffset, flags, (void *)cfdumpBytecodePrintFunction, PORTLIB, "");
+			}
+			bcvd->vmStruct = vmThread;
+			j9bcv_verifyBytecodes(vm->portLibrary, NULL, userData.romClass, bcvd, userData.romMethod, userData.bytecodeOffset, &npeMsg);
+			if (NULL != npeMsg) {
+				PORT_ACCESS_FROM_VMC(vmThread);
+				j9object_t msgObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(vmThread, (U_8 *)npeMsg, strlen(npeMsg), 0);
+				if (NULL != msgObject) {
+					msgObjectRef = vmFuncs->j9jni_createLocalRef(env, msgObject);
+					if (NULL == msgObjectRef) {
```

Comment:
This should move a bit further out - if the internal string allocation fails, you should also set this. Suggest moving to after the memory free.

Reply:
Agreed.